### DISCUSSION TITLE
Is there a way to fetch types of all variables in a file?

### DISCUSSION THREAD

**Discussion by ashwinprasadme** [microsoft/pyright#3790] [Q&A] [answered]
I have a requirement where I would like to know the types of all variables found in the python script.

Is there some way to fetch this or make pyright output this as a report?

For instance, give the following script:
```python
import numpy as np
data = np.array([1.0, 2.0, 3.0])
x = "this is a string"
```
I would like to get the types of all variables as follows:
```
data: numpy.ndarray
x: str
```

Is there a way to do this? Could anyone please point me in the right direction on how to achieve this? Thank you!

**Comment by erictraut** [comment, marked as answer]
Can you provide more details about what you are trying to do with this type information? The phrase "types of all variables" is a bit ambiguous.

A variable may have a *declared type* if it has a type declaration associated with it. Based on your example above, it looks like the variables in question do not have a declared type.

```python
# Declared type of `x` is `str`
x: str = "hi"
```

If a variable does not have a declared type, then its type is *inferred* based on all assignments to that variable from all local execution contexts.

```python
class Foo:
    def method1(self):
        self.x = "hi"

    def method2(self):
        self.x = 0

reveal_type(Foo().x) # str | int
```

Variables also have a *locally-narrowed type* based on assignments and conditional statements. The `reveal_type` call emits the locally-narrowed type.

```python
if rand() > 0.5:
    x = "hi"
    reveal_type(x) # Literal["hi"]
else:
    x = [0, 2, 3]
    reveal_type(x) # list[int]

reveal_type(x) # Literal["hi"] | list[int]
```

**Reply by ashwinprasadme** [reply]
Indeed, locally-narrowed type (aka flow-sensitive types?) is what I want for my analyzer.

I tested and see that `reveal_type` is helpful indeed. However, for my use case, I want to analyze entire files and get type information for variables at particular line numbers.

Can this be done without instrumenting the source code with `reveal_type`? How should I go about doing this? Thanks for the quick reply!

**Reply by erictraut** [reply]
In your use case, can we assume that the variable of interest is already referenced on the line where you want its type to be computed? Or do you want to be able to ask "what is the type of variable `x` on line 5" even if `x` doesn't appear anywhere on line 5?

If the variable is already referenced within the code at the point where you want to evaluate its type, you could use the LSP hover provider interface to query for the type. Since pyright is a language server, it exposes LSP functionality including the "onHover" callback. Using this interface, you can ask it "for this position in the file (line & character offset), what is the type?". What you get back is a bit of text (in plaintext or markup format). This is what you see displayed within your editor when you hover over the variable.

Does that meet your needs?

**Reply by ashwinprasadme** [reply]
That sounds like something I can work with! Thank you for the input. I will look into it!
